docs/source/cpp/gandiva.rst

```diff
@@ -0,0 +1,151 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+.. cpp:namespace:: arrow::compute
+
+===============================
+The Gandiva Expression Compiler
+===============================
+
+Gandiva is a runtime expression compiler that uses `LLVM`_ to generate
+efficient native code for projections and filters on Arrow record batches.
+Gandiva only handles projections and filters. For other transformations, see
+:ref:`Compute Functions <compute-cpp>`.
+
+Gandiva was designed to take advantage of the Arrow memory format and modern
+hardware. Compiling expressions using LLVM allows the execution to be optimized
```

Comment:
This reads oddly, because the first sentence talks about "memory format and modern hardware," then only modern hardware is discussed. Could it be rephrased to include how it leverages the memory, or just focus on the specifics of how LLVM is helpful here?

Reply:
Sure, here's one example from [this blog post](https://www.dremio.com/blog/announcing-gandiva-initiative-for-apache-arrow/). What do you think of that?

```suggestion
hardware. For example, since Arrow arrays have separate buffers for values and 
validity bitmaps, values and their null status can often be processed 
independently, allowing for better instruction pipelining. In addition,
compiling expressions using LLVM allows the execution to be optimized
```